components/bl-treeSelect-component/src/helpers.js

```diff
@@ -1,105 +1,153 @@
-      }
-    });
-  }
+function updateParentNodes(node, selectedNodes, expandedKeysMap, selectionMode) {
+  node.parents.forEach(parent => {
+    expandedKeysMap[parent] = true;
 
-  if (node.children) {
-    let count = 0;
-
-    for (const childNode of node.children) {
-      if (selectedNodes[childNode.key]?.checked) {
-        count++;
-      }
+    if (selectionMode === SelectionMode.CHECKBOX && !selectedNodes[parent]) {
+      selectedNodes[parent] = SelectedNodesState.PARTIAL;
+    }
+  });
+}
 
-      if (selectedNodes[node.key]?.checked) {
-        selectedNodes[childNode.key] = { checked: true, partialChecked: false };
-      }
+function updateChildNodes(node, selectedNodes, expandedKeysMap, selectionMode) {
+  let count = 0;
 
-      findNodesInCheckboxMode(childNode, node, selectedNodes, expandedKeysState, keysMap);
+  for (const childNode of node.children) {
+    if (selectedNodes[childNode.key]?.checked) {
+      count++;
     }
 
-    if (count === Object.keys(node.children).length) {
-      selectedNodes[node.key] = { checked: true, partialChecked: false };
+    if (selectedNodes[node.key]?.checked) {
+      selectedNodes[childNode.key] = SelectedNodesState.FULL;
     }
+
+    updateNode(childNode, node, selectedNodes, expandedKeysMap, selectionMode);
+  }
+
+  if (count === Object.keys(node.children).length) {
+    selectedNodes[node.key] = SelectedNodesState.FULL;
   }
 }
 
 export function validateSelectedNodeKeys(selectionMode, selectedNodeKeys, keysMap) {
-  const keys = selectionMode === SelectionMode.single ? { [selectedNodeKeys]: true } : { ...selectedNodeKeys };
+  const keys = selectionMode === SelectionMode.SINGLE ? { [selectedNodeKeys]: true } : { ...selectedNodeKeys };
 
   Object.keys(keys).forEach(key => {
     if (!keysMap[key]) {
-      console.error('Can't select a non existed option with key:`${ key }`.');
+      console.error(`Can't select a non existed option with key: ${ key }.`);
     }
   });
 }
+
+export function getSelectedItems(selectedNodeKeys, nodes, selectionMode) {
+  if (!selectedNodeKeys || !Object.keys(selectedNodeKeys).length) {
+    return null;
+  }
+
+  const selectedItems = selectionMode === SelectionMode.SINGLE ? { [selectedNodeKeys]: true } : { ...selectedNodeKeys };
+
+  for (const childNode of nodes) {
+    handleNode(childNode, selectedItems, selectionMode);
+  }
+
+  return selectedItems;
+}
+
+function handleNode(node, selectedItems, selectionMode) {
+  if (selectedItems[node.key]) {
+    updateSelectedItems(node, selectedItems, selectionMode);
+  }
+
+  if (!node.children) {
+    return;
+  }
+
+  for (const childNode of node.children) {
+    handleNode(childNode, selectedItems, selectionMode);
+  }
+}
+
+function updateSelectedItems(node, selectedItems, selectionMode) {
+  const { label, data } = node;
+
+  if (selectionMode === SelectionMode.CHECKBOX) {
+    selectedItems[node.key].data = data;
+    selectedItems[node.key].label = label;
```

Comment:
```suggestion
    selectedItems[node.key] = {... selectedItems[node.key], data, label };
```

But I don't really understand difference between if and else here

Reply:
```
  if (selectionMode === SelectionMode.CHECKBOX) {
    // selectedItems[node.key] == { checked: true, partialChecked: false }
    selectedItems[node.key] = { ...selectedItems[node.key], data, label };
  } else {
    // selectedItems[node.key] == true
    selectedItems[node.key] = { data, label };
  }
```